### PR TITLE
CRTX-81731: Support rendering SLA field in timer cell

### PR DESCRIPTION
**Pull request by lmichalevich**
#### Description

We want SLA fields to look as close as they can to the web-client.

Screenshot 2023-05-29 at 19 32 02

---------

#### Generated JSON
```
[
  {
    "type": "table",
    "data": [
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "1",
        "Name": "Test",
        "Remediation SLA": "Status: idle"
      },
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "3",
        "Name": "test 3",
        "Remediation SLA": "Status: running ()"
      },
      {
        "Containment SLA": "Status: ended ()",
        "Detection SLA": "Status: running ()",
        "ID": "2",
        "Name": "test",
        "Remediation SLA": "Status: idle"
      }
    ],
    "layout": {
      "classes": "striped stackable small compact",
      "columnPos": 0,
      "forceRangeMessage": "",
      "h": 3,
      "i": "dd2917b0-fd67-11ed-9bb3-6bd79a4877db",
      "readableHeaders": {
        "Containment SLA": "Containment SLA",
        "Detection SLA": "Detection SLA",
        "Remediation SLA": "Remediation SLA",
        "id": "ID",
        "name": "Name"
      },
      "reflectDimensions": true,
      "rowPos": 0,
      "tableColumns": [
        "id",
        "name",
        "Detection SLA",
        "Containment SLA",
        "Remediation SLA"
      ],
      "tableColumnsMetaData": [
        {
          "displayed": true,
          "isDefault": true,
          "key": "id",
          "width": 110
        },
        {
          "displayed": true,
          "isDefault": true,
          "key": "name",
          "width": 100
        },
        {
          "displayed": true,
          "key": "Detection SLA",
          "width": 200
        },
        {
          "displayed": true,
          "key": "Containment SLA",
          "width": 163
        },
        {
          "displayed": true,
          "key": "Remediation SLA",
          "width": 200
        }
      ],
      "w": 12
    },
    "query": {
      "type": "incident",
      "filter": {
        "query": "",
        "period": {
          "by": "",
          "byTo": "",
          "byFrom": "days",
          "toValue": null,
          "fromValue": 7,
          "field": ""
        },
        "fromDate": "0001-01-01T00:00:00Z",
        "toDate": "0001-01-01T00:00:00Z"
      }
    },
    "automation": {
      "name": "",
      "id": "",
      "args": null,
      "noEvent": false
    },
    "fromDate": "2023-05-22T16:39:57Z",
    "title": "Test",
    "emptyNotification": "No results found",
    "titleStyle": null,
    "autoPageBreak": true
  }
]
```

#### Generated Report - Before

[Report.pdf](https://github.com/demisto/sane-reports/files/11593597/incidentDailyReportTempalte.pdf)

#### Generated Report - After

[Report.pdf](https://github.com/demisto/sane-reports/files/11593604/incidentDailyReportTempalte.pdf)

------------- 

#### In this PR

Adding new field to the generated JSON named `extraData`.

#### BE changes

In order to see these changes we need to add the `extraData` section to the generated json.
it is an array of objects. each object contains the relevant information to render the time cell (see the example in the description). We also need to add inside each item the type of the cell. (type: 'timer').

The information comes from the following apis:

Screenshot 2023-05-30 at 13 11 23

threshold:

Screenshot 2023-05-30 at 13 15 49

#### Example JSON
```
[
  {
    "type": "table",
    "data": [
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "1",
        "Name": "Test",
        "Remediation SLA": "Status: idle"
      },
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "3",
        "Name": "test 3",
        "Remediation SLA": "Status: running ()"
      },
      {
        "Containment SLA": "Status: ended ()",
        "Detection SLA": "Status: running ()",
        "ID": "2",
        "Name": "test",
        "Remediation SLA": "Status: idle"
      }
    ],
    "layout": {
      "classes": "striped stackable small compact",
      "columnPos": 0,
      "forceRangeMessage": "",
      "h": 3,
      "i": "dd2917b0-fd67-11ed-9bb3-6bd79a4877db",
      "readableHeaders": {
        "Containment SLA": "Containment SLA",
        "Detection SLA": "Detection SLA",
        "Remediation SLA": "Remediation SLA",
        "id": "ID",
        "name": "Name"
      },
      "reflectDimensions": true,
      "rowPos": 0,
      "tableColumns": [
        "id",
        "name",
        "Detection SLA",
        "Containment SLA",
        "Remediation SLA"
      ],
      "tableColumnsMetaData": [
        {
          "displayed": true,
          "isDefault": true,
          "key": "id",
          "width": 110
        },
        {
          "displayed": true,
          "isDefault": true,
          "key": "name",
          "width": 100
        },
        {
          "displayed": true,
          "key": "Detection SLA",
          "width": 200
        },
        {
          "displayed": true,
          "key": "Containment SLA",
          "width": 163
        },
        {
          "displayed": true,
          "key": "Remediation SLA",
          "width": 200
        }
      ],
      "w": 12
    },
    "query": {
      "type": "incident",
      "filter": {
        "query": "",
        "period": {
          "by": "",
          "byTo": "",
          "byFrom": "days",
          "toValue": null,
          "fromValue": 7,
          "field": ""
        },
        "fromDate": "0001-01-01T00:00:00Z",
        "toDate": "0001-01-01T00:00:00Z"
      }
    },
    "extraData": [
      {
        "Detection SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-28T18:34:26.447968+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 20,
          "slaStatus": -1,
          "startDate": "2023-05-28T18:14:26.447968+03:00",
          "totalDuration": 0,
          "threshold": 0
        },
        "Containment SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 30,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 3
        },
        "Remediation SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 7200,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 0
        }
      },
      {
        "Detection SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-28T18:18:40.697968+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 2,
          "slaStatus": -1,
          "startDate": "2023-05-28T18:16:40.697968+03:00",
          "totalDuration": 0,
          "threshold": 0
        },
        "Containment SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 30,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 3
        },
        "Remediation SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-06-03T19:28:32.280921+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 7200,
          "slaStatus": -1,
          "startDate": "2023-05-29T19:28:32.280921+03:00",
          "totalDuration": 0,
          "threshold": 0
        }
      },
      {
        "Detection SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-28T18:15:28.339413+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 20,
          "slaStatus": -1,
          "startDate": "2023-05-28T17:55:28.339413+03:00",
          "totalDuration": 0,
          "threshold": 0
        },
        "Containment SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-29T19:56:29.235357+03:00",
          "endDate": "2023-05-29T19:29:31.44581+03:00",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "ended",
          "sla": 30,
          "slaStatus": 1,
          "startDate": "2023-05-29T19:26:29.235357+03:00",
          "totalDuration": 182,
          "threshold": 3
        },
        "Remediation SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 7200,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 0
        }
      }
    ],
    "automation": {
      "name": "",
      "id": "",
      "args": null,
      "noEvent": false
    },
    "fromDate": "2023-05-22T16:39:57Z",
    "title": "Test",
    "emptyNotification": "No results found",
    "titleStyle": null,
    "autoPageBreak": true
  }
]
```

#### Generated Report - Before

[Report.pdf](https://github.com/demisto/sane-reports/files/11593590/incidentDailyReportTempalte.pdf)

#### Generated Report - After

[Report.pdf](https://github.com/demisto/sane-reports/files/11593663/incidentDailyReportTempalte.pdf)

### CODE REVIEW

**Comment by yanyo87** [comment]
image
remove these files

**Comment by lmichalevich** [comment]
Tested on another JSON:

[test.json](https://github.com/demisto/sane-reports/files/11599488/test.txt)

[after.pdf](https://github.com/demisto/sane-reports/files/11599466/after.pdf)
[before.pdf](https://github.com/demisto/sane-reports/files/11599470/before.pdf)

File: src/css/index.less [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+i {
+  &.demisto-icon {
+    font-size: 24px !important;
+    display: inline-block;
+    margin: 0 !important;
+    vertical-align: sub !important;
+  }
+}
```
Comment (yanyo87) [inline src/css/index.less:13]: don't do this globally,
add the relevant part only to the timer cell

File: src/utils/time.js [hunk 2]
```diff
@@ -31,3 +31,95 @@ export function isDateNotValid(d) {
...
+
+export function getSlaProps(timerValue) {
+  const shouldTick = timerValue.runStatus === TIMER_STATUS.running;
+  const timePeriod = EXTENDED_TIME_PERIODS.SECONDS;
+  const duration = getSlaDurationOrNull(timerValue);
+  const dueDate = timerValue.dueDate;
+  const startDate = getSlaEffectiveStartTime(timerValue);
+  const riskThreshold = timerValue.threshold || 0;
+
+  return { shouldTick, timePeriod, duration, startDate, dueDate, riskThreshold };
```
Comment (yanyo87) [inline src/utils/time.js:124]:
```suggestion

  const duration = getSlaDurationOrNull(timerValue);
  const dueDate = timerValue.dueDate;
  const startDate = getSlaEffectiveStartTime(timerValue);
  const riskThreshold = timerValue.threshold || 0;

  return { shouldTick: timerValue.runStatus === TIMER_STATUS.running, timePeriod :EXTENDED_TIME_PERIODS.SECONDS , duration ...., startDate, dueDate, riskThreshold };
```

File: src/utils/time.js [hunk 3]
```diff
@@ -31,3 +31,95 @@ export function isDateNotValid(d) {
...
+  const startDate = getSlaEffectiveStartTime(timerValue);
+  const stopDate = getSlaEffectiveStopDate(timerValue);
+
+  return moment(stopDate).diff(startDate, EXTENDED_TIME_PERIODS.SECONDS);
+}
+
+function getSlaTimeLeft(timerValue) {
+  const effectiveStopDate = getSlaEffectiveStopDate(timerValue);
+
+  return moment(timerValue.dueDate).diff(effectiveStopDate, EXTENDED_TIME_PERIODS.SECONDS);
```
Comment (yanyo87) [inline src/utils/time.js:103]:
```suggestion
  return moment(timerValue.dueDate).diff(getSlaEffectiveStopDate(timerValue), EXTENDED_TIME_PERIODS.SECONDS);
```

Comment (lmichalevich) [reply src/utils/time.js:103]: It is like this in the web-client. I think both of the way to right code are ok (no preferred way).

File: src/css/index.less (hunk 1 above)
Comment (lmichalevich) [reply src/css/index.less]: moved them to the timer cell

**Comment by lmichalevich** [comment]
Updated the reports.

Added the SLA time with ellipsis (the format is too long and otherwise it looks bad when it breaks)

#### Before

[incidentDailyReportTempalte.pdf](https://github.com/demisto/sane-reports/files/11674851/incidentDailyReportTempalte.pdf)

#### After

[incidentDailyReportTempalte.pdf](https://github.com/demisto/sane-reports/files/11674834/incidentDailyReportTempalte.pdf)

**Review by yanyo87 — APPROVED** [review]

**Pull request merged**